pkg/download/buffered_reader.go

```diff
@@ -39,6 +46,16 @@ func newBufferedReader(pool *bufferPool) *bufferedReader {
 // pool.
 func (b *bufferedReader) Read(buf []byte) (int, error) {
 	<-b.ready
+	err := b.readErr()
+	if err != nil {
+		return 0, err
+	}
+	if b.buf == uninitializedReader {
```

Comment:
More of a nit than meaningful feedback, but using a sentinel `bufio.Reader` for this case seems a bit clever. It also allocates a default-size buffer that is never used.

Perhaps you could define the `buf` field in terms of the interface you require rather than as a concrete `*bufio.Reader`. That would allow you to treat `nil` as uninitialized, and when you're done with it you can replace it with your own `alwaysReturnEOFReader{}`?

Reply:
hmm, I think I can just use nil anyway since *bufio.Reader is a pointer type